### PR TITLE
Active Account, Password Reset

### PR DESCRIPTION
**Pull request by TuanNhon**

### CODE REVIEW

**Comment by TuanNhon** [comment]
ready

File: app/controllers/account_activations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class AccountActivationsController < ApplicationController
+
```
Comment (huuphuc93) [inline app/controllers/account_activations_controller.rb]: xóa dòng trắng

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  def edit; end
+
+  private
+
+  def user_params
+    params.require(:user).permit :password, :password_confirmation
+  end
+
+  def get_user
+    @user = User.find_by email: params[:email]
```
Comment (huuphuc93) [inline app/controllers/password_resets_controller.rb]: check đk k tìm thấy user

File: app/models/user.rb [hunk 3]
```diff
@@ -30,12 +33,46 @@ def remember
...
+    update_attributes reset_digest: User.digest(reset_token),
+      reset_sent_at: Time.zone.now
+  end
+
+  def send_password_reset_email
+    UserMailer.password_reset(self).deliver_now
+  end
+
+  def password_reset_expired?
+    reset_sent_at < 2.hours.ago
```
Comment (huuphuc93) [inline app/models/user.rb]: chuyển 2 vào Settings

**Comment by huuphuc93** [comment]
commented

**Comment by TuanNhon** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    render :new
+  end
+
+  def valid_user
+    return if @user&.activated? && @user.authenticated?(:reset, params[:id])
+    redirect_to root_url
+  end
+
+  def check_expiration
+    if @user.password_reset_expired?
```
Comment (huuphuc93) [inline app/controllers/password_resets_controller.rb]: dùng return nhé

**Comment by huuphuc93** [comment]
commented 
comment nhỏ

**Comment by TuanNhon** [comment]
ready

**Comment by huuphuc93** [comment]
merged
this (pr) can be (merged)  now
CC: [To:2076564] Nguyen Huy Hung C (93)
CC: [To:1818111] Nguyen Thi Huong (94)
CC: [To:1546746] Md. Mahmudur Rahman (Mahmud)

File: app/controllers/password_resets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+  def edit; end
+
+  private
+
+  def user_params
+    params.require(:user).permit :password, :password_confirmation
+  end
+
+  def get_user
```
Comment (mdmahmudurrahman) [inline app/controllers/password_resets_controller.rb]: Bad naming, you should name a method in ruby using get/set keyword, you can name it like as load_user

File: app/controllers/password_resets_controller.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: %i(edit update)
+  before_action :valid_user, only: %i(edit update)
+  before_action :check_expiration, only: %i(edit update)
```
Comment (mdmahmudurrahman) [inline app/controllers/password_resets_controller.rb]: As the actions are same, you can combine line 2, 3 and 4 into one

File: config/environments/development.rb [hunk 7]
```diff
@@ -1,61 +1,38 @@
...
-  # Use an evented file watcher to asynchronously detect changes in source code,
-  # routes, locales, etc. This feature depends on the listen gem.
   config.file_watcher = ActiveSupport::EventedFileUpdateChecker
+  config.action_mailer.delivery_method = :smtp
+  config.action_mailer.smtp_settings = {
+    address: "smtp.gmail.com",
+    port: "587",
+    domain: "example.com",
+    user_name: "nhongbay2112bt@gmail.com",
+    password: "hovpuoyxugoabvgu",
```
Comment (mdmahmudurrahman) [inline config/environments/development.rb]: Don't place private credential as a plain text to this file, you can use environment variable/figaro gem

**Comment by mdmahmudurrahman** [comment]
commented

**Comment by TuanNhon** [comment]
ready

**Pull request merged**